// @sanity/desk-tool/lib/panes/list/ListPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListPane = ListPane;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _icons = require("@sanity/icons");

var _components = require("../../components");

var _paneItem = require("../../components/paneItem");

var _deskTool = require("../../contexts/deskTool");

var _paneRouter = require("../../contexts/paneRouter");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var Divider = _styledComponents.default.hr(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  background-color: var(--card-border-color);\n  height: 1px;\n  margin: 0;\n  border: none;\n"])));
/**
 * @internal
 */


function ListPane(props) {
  var childItemId = props.childItemId,
      index = props.index,
      isSelected = props.isSelected,
      isActive = props.isActive,
      pane = props.pane,
      paneKey = props.paneKey;

  var _useDeskTool = (0, _deskTool.useDeskTool)(),
      features = _useDeskTool.features;

  var _usePaneLayout = (0, _components.usePaneLayout)(),
      layoutCollapsed = _usePaneLayout.collapsed;

  var defaultLayout = pane.defaultLayout,
      _pane$displayOptions = pane.displayOptions,
      displayOptions = _pane$displayOptions === void 0 ? {} : _pane$displayOptions,
      _pane$items = pane.items,
      items = _pane$items === void 0 ? [] : _pane$items,
      _pane$menuItems = pane.menuItems,
      menuItems = _pane$menuItems === void 0 ? [] : _pane$menuItems,
      _pane$menuItemGroups = pane.menuItemGroups,
      menuItemGroups = _pane$menuItemGroups === void 0 ? [] : _pane$menuItemGroups,
      title = pane.title;
  var paneShowIcons = displayOptions.showIcons;
  var itemIsSelected = (0, _react.useCallback)(item => {
    return childItemId === item.id;
  }, [childItemId]);
  var shouldShowIconForItem = (0, _react.useCallback)(item => {
    var _item$displayOptions;

    var itemShowIcon = (_item$displayOptions = item.displayOptions) === null || _item$displayOptions === void 0 ? void 0 : _item$displayOptions.showIcon; // Specific true/false on item should have presedence over list setting

    if (typeof itemShowIcon !== 'undefined') {
      return itemShowIcon !== false; // Boolean(item.icon)
    } // If no item setting is defined, defer to the pane settings


    return paneShowIcons !== false; // Boolean(item.icon)
  }, [paneShowIcons]);
  var handleAction = (0, _react.useCallback)(item => {
    if (typeof item.action === 'function') {
      item.action(item.params);
      return;
    }

    if (typeof item.action === 'string') {
      // eslint-disable-next-line no-console
      console.warn('No handler for action:', item.action);
      return;
    } // eslint-disable-next-line no-console


    console.warn('The menu item is missing the `action` property');
  }, []);
  var actions = (0, _react.useMemo)(() => menuItems.length > 0 && /*#__PURE__*/_react.default.createElement(_components.PaneContextMenuButton, {
    items: menuItems,
    itemGroups: menuItemGroups,
    onAction: handleAction
  }), [handleAction, menuItemGroups, menuItems]);
  var header = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_components.PaneHeader, {
    actions: actions,
    backButton: features.backButton && index > 0 && /*#__PURE__*/_react.default.createElement(_ui.Button, {
      as: _paneRouter.BackLink,
      "data-as": "a",
      icon: _icons.ArrowLeftIcon,
      mode: "bleed"
    }),
    title: title
  }), [actions, features.backButton, index, title]);
  var content = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_components.PaneContent, {
    overflow: layoutCollapsed ? undefined : 'auto'
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    padding: 2,
    space: 1
  }, items.map(item => {
    if (item.type === 'divider') {
      return /*#__PURE__*/_react.default.createElement(_ui.Box, {
        key: item.id,
        paddingY: 1
      }, /*#__PURE__*/_react.default.createElement(Divider, null));
    }

    return /*#__PURE__*/_react.default.createElement(_paneItem.PaneItem, {
      icon: shouldShowIconForItem(item) ? item.icon : false,
      id: item.id,
      isActive: isActive,
      isSelected: itemIsSelected(item),
      key: item.id,
      layout: defaultLayout,
      schemaType: item.schemaType,
      title: item.title
    });
  }))), [defaultLayout, isActive, itemIsSelected, items, layoutCollapsed, shouldShowIconForItem]);
  return (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_components.Pane, {
    currentMaxWidth: 350,
    "data-index": index,
    "data-pane-key": paneKey,
    "data-testid": "desk-tool-list-pane",
    maxWidth: 640,
    minWidth: 320,
    selected: isSelected
  }, header, content), [content, header, index, isSelected, paneKey]);
}
