// @sanity/default-layout/lib/navbar/studioStatus/UpdateNotifierDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _VersionsTable = require("./VersionsTable");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var upperFirst = str => "".concat(str.slice(0, 1).toUpperCase()).concat(str.slice(1));

class UpdateNotifierDialog extends _react.default.PureComponent {
  renderTable() {
    var outdated = this.props.outdated;
    var rows = outdated.map(pkg => {
      return {
        name: pkg.name,
        items: [pkg.version, pkg.latest, upperFirst(pkg.severity || 'low')]
      };
    });
    return /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 5
    }, /*#__PURE__*/_react.default.createElement(_VersionsTable.VersionsTable, {
      headings: ['Module', 'Installed', 'Latest', 'Importance'],
      rows: rows
    }), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "To upgrade, run the ", /*#__PURE__*/_react.default.createElement("a", {
      href: "https://www.sanity.io/docs/reference/cli"
    }, "Sanity CLI"), ' ', "upgrade command in your project folder from a terminal."), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
      space: 3,
      paddingLeft: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true
    }, /*#__PURE__*/_react.default.createElement(_icons.TerminalIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Code, null, "sanity upgrade"))));
  }

  renderContactDeveloper() {
    var severity = this.props.severity;
    return /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, severity === 'high' ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "This Studio should be updated. Please get in touch with the developers and ask them to upgrade it for you.") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "This Studio has available upgrades. Consider getting in touch with the developers and ask them to upgrade it for you.")), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      as: "details",
      paddingTop: 4,
      borderTop: true
    }, /*#__PURE__*/_react.default.createElement("summary", null, "Developer info"), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 4
    }, this.renderTable())));
  }

  render() {
    var _this$props = this.props,
        severity = _this$props.severity,
        onClose = _this$props.onClose;
    return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
      zOffset: "navbarDialog"
    }, /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
      header: severity === 'low' ? 'Upgrades available' : 'Studio is outdated',
      onClose: onClose,
      onClickOutside: onClose,
      id: "update-notifier-dialog",
      width: 1,
      scheme: "light"
    }, __DEV__ && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingY: 5,
      paddingX: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "This Studio is no longer up to date", ' ', severity === 'high' ? 'and should be upgraded.' : 'and can be upgraded.'), this.renderTable())), !__DEV__ && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingY: 5,
      paddingX: 4
    }, this.renderContactDeveloper())));
  }

}

_defineProperty(UpdateNotifierDialog, "defaultProps", {
  outdated: []
});

var _default = UpdateNotifierDialog;
exports.default = _default;
