// @sanity/desk-tool/lib/components/pane/__workshop__/SplitPanes/Navbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Navbar = Navbar;

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var RootCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  padding-left: env(safe-area-inset-left);\n  padding-right: env(safe-area-inset-right);\n"])));
var features = {
  inverted: true
};

function Navbar() {
  var theme = (0, _ui.useTheme)();
  var dark = theme.sanity.color.dark;
  return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "navbar"
  }, /*#__PURE__*/_react.default.createElement(_ui.Layer, null, /*#__PURE__*/_react.default.createElement(RootCard, {
    scheme: features.inverted ? 'dark' : undefined,
    shadow: features.inverted ? dark ? 1 : 0 : 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "bold"
  }, "Studio")))));
}
