// @sanity/desk-tool/lib/panes/document/documentPanel/header/DocumentHeaderTitle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentHeaderTitle = DocumentHeaderTitle;

var _react = _interopRequireDefault(require("react"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _preview = require("part:@sanity/base/preview");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var PREVIEW_FIELDS = ['title'];

function renderTitle(_ref) {
  var title = _ref.title;
  return title ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, title) : /*#__PURE__*/_react.default.createElement("em", null, "Untitled");
}

function DocumentHeaderTitle(props) {
  var documentType = props.documentType,
      paneTitle = props.paneTitle,
      value = props.value;

  var type = _schema.default.get(documentType);

  if (paneTitle) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, paneTitle);
  }

  if (!value) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "New ", (type === null || type === void 0 ? void 0 : type.title) || (type === null || type === void 0 ? void 0 : type.name));
  }

  return /*#__PURE__*/_react.default.createElement(_preview.PreviewFields, {
    document: value,
    layout: "inline",
    type: type,
    fields: PREVIEW_FIELDS
  }, renderTitle);
}
